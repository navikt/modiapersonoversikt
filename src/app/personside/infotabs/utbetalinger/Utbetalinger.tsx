import * as React from 'react';
import { Utbetaling, UtbetalingerResponse } from '../../../../models/utbetalinger';
import styled from 'styled-components';
import theme from '../../../../styles/personOversiktTheme';
import { Undertittel } from 'nav-frontend-typografi';
import { FilterState } from './filter/Filter';
import TotaltUtbetalt from './totalt utbetalt/TotaltUtbetalt';
import { ArrayGroup, groupArray, GroupedArray } from '../../../../utils/groupArray';
import {
    fjernTommeUtbetalinger,
    getTypeFromYtelse,
    månedOgÅrForUtbetaling,
    reduceUtbetlingerTilYtelser,
    utbetalingDatoComparator,
    utbetaltTilBruker
} from './utils/utbetalingerUtils';
import { AlertStripeInfo } from 'nav-frontend-alertstriper';
import Månedsgruppe from './MånedsGruppe';
import HandleUtbetalingerArrowKeys from './HandleUtbetalingerHotKeys';

const UtbetalingerArticle = styled.article`
    ${theme.hvittPanel};
    margin-top: ${theme.margin.layout};
    margin-bottom: ${theme.margin.layout};
    > *:first-child {
        padding: ${theme.margin.px20};
    }
`;

const UtbetalingerListe = styled.ol`
    padding: 0;
    margin: 0;
`;

const Wrapper = styled.div`
    ol {
        list-style: none;
    }
`;

export function getFiltrerteUtbetalinger(utbetalinger: Utbetaling[], filter: FilterState) {
    return utbetalinger
        .filter(utbetaling => filtrerPaUtbetaltTilValg(utbetaling, filter))
        .map(utbetaling => filtrerBortYtelserSomIkkeErValgt(utbetaling, filter))
        .filter(fjernTommeUtbetalinger);
}

function filtrerPaUtbetaltTilValg(utbetaling: Utbetaling, filter: FilterState) {
    return filter.utbetaltTil.includes(utbetaling.erUtbetaltTilPerson ? utbetaltTilBruker : utbetaling.utbetaltTil);
}

function filtrerBortYtelserSomIkkeErValgt(utbetaling: Utbetaling, filter: FilterState): Utbetaling {
    const ytelser = reduceUtbetlingerTilYtelser([utbetaling]).filter(ytelse =>
        filter.ytelser.includes(getTypeFromYtelse(ytelse))
    );
    return {
        ...utbetaling,
        ytelser: ytelser
    };
}

interface UtbetalingerProps {
    utbetalingerData: UtbetalingerResponse;
    filter: FilterState;
}

function Utbetalinger({ filter, ...props }: UtbetalingerProps) {
    const filtrerteUtbetalinger = getFiltrerteUtbetalinger(props.utbetalingerData.utbetalinger, filter);
    if (filtrerteUtbetalinger.length === 0) {
        return (
            <AlertStripeInfo>
                Det finnes ingen utbetalinger for valgte kombinasjon av periode og filtrering.
            </AlertStripeInfo>
        );
    }

    const utbetalingerGruppert: GroupedArray<Utbetaling> = groupArray(
        filtrerteUtbetalinger.sort(utbetalingDatoComparator),
        månedOgÅrForUtbetaling
    );

    const månedsGrupper = utbetalingerGruppert.map((gruppe: ArrayGroup<Utbetaling>) => (
        <Månedsgruppe key={gruppe.category} gruppe={gruppe} {...props} />
    ));

    return (
        <Wrapper>
            <TotaltUtbetalt utbetalinger={filtrerteUtbetalinger} periode={props.utbetalingerData.periode} />
            <HandleUtbetalingerArrowKeys utbetalinger={filtrerteUtbetalinger}>
                <UtbetalingerArticle aria-label="Utbetalinger">
                    <Undertittel>Utbetalinger</Undertittel>
                    <UtbetalingerListe aria-label="Måneder">{månedsGrupper}</UtbetalingerListe>
                </UtbetalingerArticle>
            </HandleUtbetalingerArrowKeys>
        </Wrapper>
    );
}

export default Utbetalinger;
