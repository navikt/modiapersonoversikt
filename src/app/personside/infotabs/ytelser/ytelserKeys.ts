// Brukes til å finne ut hvilke utbetalinger fra UtbetalingerRestkonto som tilhører disse ytelsene, må matche kodeverk
export enum YtelserKeys {
    Foreldrepenger = 'Foreldrepenger',
    Sykepenger = 'Sykepenger'
}
