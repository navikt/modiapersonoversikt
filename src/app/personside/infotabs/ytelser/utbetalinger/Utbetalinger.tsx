import * as React from 'react';
import { HistoriskUtbetaling, KommendeUtbetaling } from '../../../../../models/ytelse/ytelse-utbetalinger';
import styled from 'styled-components';
import theme from '../../../../../styles/personOversiktTheme';
import KommendeUtbetalinger from './KommendeUtbetalinger';
import { YtelserKeys } from '../ytelserKeys';
import HistoriskeUtbetalingerContainer from './HistoriskeUtbetalingerContainer';
import ErrorBoundary from '../../../../../components/ErrorBoundary';

interface Props {
    kommendeUtbetalinger: KommendeUtbetaling[];
    historiskeUtbetalinger: HistoriskUtbetaling[];
    ytelsesType: YtelserKeys;
}

const StyledSection = styled.section`
    > * {
        margin: ${theme.margin.px10} 0;
    }
    margin-right: 1rem;
`;

function Utbetalinger(props: Props) {
    return (
        <ErrorBoundary boundaryName="Utbetalinger Ytelser">
            <StyledSection aria-label={'Utbetalinger ' + props.ytelsesType}>
                <KommendeUtbetalinger kommendeUtbetalinger={props.kommendeUtbetalinger} />
                <HistoriskeUtbetalingerContainer
                    historiskeUtbetalinger={props.historiskeUtbetalinger}
                    ytelseType={props.ytelsesType}
                />
            </StyledSection>
        </ErrorBoundary>
    );
}

export default Utbetalinger;
