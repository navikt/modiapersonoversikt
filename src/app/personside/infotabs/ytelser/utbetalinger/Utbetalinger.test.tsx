import * as React from 'react';
import * as renderer from 'react-test-renderer';
import TestProvider from '../../../../../test/Testprovider';
import Utbetalinger from './Utbetalinger';
import { YtelserKeys } from '../ytelserKeys';

test('Utbetalinger matcher snapshot', () => {
    const result = renderer.create(
        <TestProvider>
            <Utbetalinger
                kommendeUtbetalinger={[]}
                historiskeUtbetalinger={[]}
                ytelsesType={YtelserKeys.Foreldrepenger}
            />
        </TestProvider>
    );

    expect(result).toMatchSnapshot();
});
