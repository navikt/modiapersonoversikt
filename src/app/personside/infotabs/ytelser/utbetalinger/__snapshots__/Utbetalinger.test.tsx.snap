// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Utbetalinger matcher snapshot 1`] = `
.c2 {
  text-align: center;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1 {
  padding: 0.5rem;
}

.c0 {
  margin-right: 1rem;
}

.c0 > * {
  margin: 0.625rem 0;
}

<section
  aria-label="Utbetalinger Foreldrepenger"
  className="c0"
>
  <div
    className="alertstripe alertstripe--info"
  >
    <span
      aria-label="info"
      className="alertstripe__ikon"
    >
      <svg
        focusable="false"
        height="1.5em"
        kind="info-sirkel-fyll"
        viewBox="0 0 24 24"
        width="1.5em"
      >
        <g
          fill="none"
          fillRule="evenodd"
        >
          <path
            d="M12.205-.004l-.214.002a12.225 12.225 0 0 0-8.517 3.659C1.179 5.977-.053 9.013.002 12.208c.115 6.613 5.296 11.793 11.795 11.793l.212-.002c6.726-.116 12.105-5.595 11.99-12.21C23.883 5.178 18.702-.003 12.204-.003z"
            fill="#FFA733"
            fillRule="nonzero"
          />
          <path
            d="M12.027 19H12A1.499 1.499 0 0 1 11.973 16L12 16a1.501 1.501 0 0 1 .027 3z"
            fill="#3E3832"
          />
          <path
            d="M12 5a1 1 0 0 1 1 1v7a1 1 0 0 1-2 0V6a1 1 0 0 1 1-1z"
            fill="#3E3832"
            fillRule="nonzero"
          />
        </g>
      </svg>
    </span>
    <span
      className="typo-normal alertstripe__tekst"
    >
      Ingen kommende utbetalinger funnet
    </span>
  </div>
  <section>
    <div
      className="c1"
    >
      <div
        className="c2"
      >
        <h4
          className="typo-undertittel"
        >
          UtfÃ¸rte utbetalinger
        </h4>
      </div>
    </div>
    <ol />
    <div
      className="c3"
    >
      <button
        className="knapp knapp--standard"
        disabled={false}
        onClick={[Function]}
        type="submit"
      >
        Hent alle historiske utbetalinger
      </button>
    </div>
  </section>
</section>
`;
